### ISSUE TITLE
cannot load such file -- bundler/dep_proxy & Bundler 1.16.3

### ISSUE THREAD

**Issue by mattheworiordan** [heroku/heroku-buildpack-ruby#802]
@schneems you asked for a fresh issue, here goes in relation to https://github.com/heroku/heroku-buildpack-ruby/issues/655#issuecomment-400699589, which is the recently closed issue in regards  to a Bundler upgrade.

Some background on our use case. We make use of Heroku buildpacks both with Heroku itself, and within our infrastructure to build slugs that we deploy inside containers.

Since upgrading to Ruby 2.5.*+, I continue to get this failure from Bunlder (1.15.*, 1.16.1 even):

```
/slugs/foo/vendor/ruby-2.5.1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `require': cannot load such file -- bundler/dep_proxy (LoadError)
        from /slugs/foo/vendor/ruby-2.5.1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `rescue in require'
        from /slugs/foo/vendor/ruby-2.5.1/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:39:in `require'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:856:in `block (2 levels) in expand_dependencies'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:855:in `each'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:855:in `block in expand_dependencies'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:842:in `each'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:842:in `expand_dependencies'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:779:in `converge_locked_specs'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:228:in `resolve'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:159:in `specs'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/definition.rb:218:in `specs_for'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler/runtime.rb:21:in `setup'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler.rb:103:in `setup'
        from /slugs/foo/vendor/bundle/ruby/2.5.0/gems/bundler-1.15.2/lib/bundler.rb:108:in `require'
        from /manager/current/bin/health-server:21:in `<main>'
```

Specifically this happens when calling:

```ruby
Bundler.require :default
```

When I use the slug manually and run a basic Ruby console with `irb`, calling `Bundler.require :default` does the same thing.

However, if I run `gem install bundler` within that slug's environment (I set home to the extracted slug path, source the `profile/ruby.sh` file), it installs Bundler 1.16.4 just fine, and then subsequently calling `Bundler.require :default` works as well.

Interestingly though, the Gem path is: `~/vendor/bundle/ruby/2.5.0/gems`, and in there is as expected, Bundler 1.15.2 or 1.16.1 (when using my PR that was rejected https://github.com/heroku/heroku-buildpack-ruby/pull/738).

However, when running `gem install bundler`, the `bundler` gem is installed in another folder:

```
/slugs/foo/vendor/ruby-2.5.1/lib/ruby/gems/2.5.0/gems/bundler-1.16.4
```

Others have experienced similar issues with Bundler (although admittedly with TruffleRuby), and claim 1.16.3 fixes the problem.  However, when I try and use 1.16.3 in a forked version of this buildpack, it fails because of a dependency on Heroku:

```
02:58:01 WARN  /buildpacks/heroku-buildpack-ruby/lib/language_pack/shell_helpers.rb:95:in `block in run!': Command: 'set -o pipefail; curl -L --fail --retry 5 --retry-delay 1 --connect-timeout 3 --max-time 30 https://s3-external-1.amazonaws.com/heroku-buildpack-ruby/bundler-1.16.3.tgz -s -o - | tar zxf - ' failed unexpectedly: (LanguagePack::Fetcher::FetchError)
```

So I get the sense we possibly have two problems here:

* The installation of Bundler from your compiled Heroku tar file is causing issues with newer versions of Ruby for some reason
* I am unable to confirm if the issue is resolved with a newer version because there is a dependency on Heroku to have a binary tar of bundler.

I realise this issue is probably not going to be easy to reproduce, but please shout if you have any theories or advice on what I can do to work around this issue or help fix it.

**Comment by schneems** [comment]
Sorry somehow I missed this issue coming in. Can you try out the branch `schneems/bundler-1.16.4` I staged 1.16.4 and that branch should work (in theory). You can use a branch of a buildpack by running the `heroku buildpacks` command and using `https://github.com/heroku/heroku-buildpack-ruby#schneems/bundler-1.16.4` as the value.

**Comment by schneems** [comment]
Also if you can open support ticket help.heroku.com search for "bundler" and link this issue I can take a look at your app I might be able to help. I'm not seeing this on other apps so this is likely something that is triggering this behavior.

**Comment by mattheworiordan** [comment]
@schneems sorry been stacked, but will look at this very soon.

**Comment by mattheworiordan** [comment]
@schneems apologies I went off the grid on this one, but sadly as I had a workaround, I had to focus on other areas.  Now that I have some time over the festive period, I wanted to revisit this, and I found that when using your PR to [use Bundler 1.17.2](https://github.com/heroku/heroku-buildpack-ruby/pull/835), with Ruby 2.5 and Ruby 2.6, both work just fine now.

So I am closing this issue as it is no longer an issue, well at least it won't be once you add support for a newer version of Bundler into `master`.

Thanks for looking into this, and nice to see Bundler is finally going to be upgraded :)

**Comment by mattheworiordan** [comment]
@schneems as a side note, this in fact did not work with Ruby 2.5, ony Ruby 2.6.  I can help reproduce if you think it's worth investigating further. For now, we're doing an emergency upgrade to Ruby 2.6 across the board (we use Heroku build packs to builds slugs for our internal Ruby orchestration services) because we got hit badly today with a workaround we had in place that broken when Bundler 2.0 was released this morning.
